bokeh/protocol.py

```diff
@@ -68,6 +69,34 @@ def transform_numerical_array(self, obj):
             transformed[np.isneginf(obj)] = '-Infinity'
             return transformed.tolist()
 
+    def traverse_data(self, datum):
+        datum_copy = []
+        for item in datum:
+            if isinstance(item, (list, tuple)):
+                datum_copy.append(self.traverse_data(item))
+            elif isinstance(item, np.ndarray):
+                datum_copy.append(self.transform_array(item))
+            elif isinstance(item, float):
+                if np.isnan(item):
+                    item = 'NaN'
+                elif np.isposinf(item):
+                    item = 'Infinity'
+                elif np.isneginf(item):
+                    item = '-Infinity'
+                datum_copy.append(item)
+            else:
+                datum_copy.append(item)
+        return datum_copy
+
+    def transform_column_source_data(self, data):
+        data_copy = {}
+        for key in iterkeys(data):
+            if is_pandas and isinstance(data[key], (pd.Series, pd.Index)):
+                data_copy[key] = self.transform_series(data[key])
+            else:
+                data_copy[key] = self.traverse_data(data[key])
+        return data_copy
```

Comment:
I think we could factor the implementation into helper functions outside the class? Then we would not have to also instantiate a `BokehJSONEncoder` in `vm_serialize`. Also I think there needs to be a numpy array check? Otherwise I don't see where it would hit `transform_numerical_array` in the code path.